File: Python/LinkedList/CircularDoublyLinkedList.py
class Node:
    def __init__(self, data) -> None:
        self.data = data
        self.next = None
        self.prev = None


class CircularDoublyLinkedList:
    def __init__(self) -> None:
        self.head = None
        self.tail = None

    def add(self, data):
        newNode = Node(data)
        if self.head is None:
            self.head = newNode
            self.tail = newNode
            self.head.next = self.head
            self.head.prev = self.head
        else:
            newNode.prev = self.tail
            newNode.next = self.head
            self.tail.next = newNode
            self.head.prev = newNode
            self.tail = newNode

    def remove(self, data):
        if self.head is None:
            return
        current = self.head
        found = False

        while True:

            if current.data == data:
                found = True
                break
            current = current.next

            if current == self.head:
                break

        if not found:
            return
        
        
        if current == self.head and current == self.tail:
            self.head = None
            self.tail = None
        elif current == self.head:
            self.tail.next = self.head.next
            self.head.next.prev = self.tail
            self.head = self.head.next
        elif current == self.tail:
            self.tail.prev.next = self.head
            self.head.prev = self.tail.prev
            self.tail = self.tail.prev
        else:
            current.prev.next = current.next
            current.next.prev = current.prev

    def totalValue(self):
        if self.head is None:
            return 0
        count = 1
        current = self.head
        while current.next != self.head:
            count += 1
            current = current.next
        return count
    
    def showCircularDoublyLinkedList(self):
        if self.head is None:
            print("The list is empty.")
            return
        current = self.head
        while True:
            print(f"Address {id(current)}")
            print(f"Data {current.data}")
            print(f"Prev Address {id(current.prev)}")
            print(f"Next Address {id(current.next)}")
            print(" ")
            current = current.next
            if current == self.head:
                break


# Test
s = CircularDoublyLinkedList()
s.add(10)
s.add(20)
s.add(30)
s.add(35)
s.add(40)
s.remove(40)
s.showCircularDoublyLinkedList()

File: Java/LinkedList/CircularDoublyLinkedList.java
package Java.LinkedList;

class Node {
    int data;
    Node next;
    Node prev;

    Node(int data) {
        this.data = data;
        this.next = null;
        this.prev = null;
    }
}

public class CircularDoublyLinkedList {
    private Node head;
    private Node tail;

    CircularDoublyLinkedList() {
        this.head = null;
        this.tail = null;
    }

    public void add(int data) {
        Node newNode = new Node(data);
        if (head == null) {
            head = newNode;
            tail = newNode;
            head.next = head;
            head.prev = head;
        } else {
            newNode.prev = tail;
            newNode.next = head;
            tail.next = newNode;
            head.prev = newNode;
            tail = newNode;
        }
    }

    public boolean search(int data) {
        if (head == null) {
            return false;
        }
        Node current = head;
        do {
            if (current.data == data) {
                return true;
            }
            current = current.next;
        } while (current != head);
        return false;
    }

    public void delete(int data) {
        if (head == null) {
            return;
        }
        if (head.data == data) {
            if (head == tail) {
                head = null;
                tail = null;
            } else {
                head = head.next;
                head.prev = tail;
                tail.next = head;
            }
            return;
        }

        Node current = head;
        do {
            if (current.data == data) {
                current.prev.next = current.next;
                current.next.prev = current.prev;
                if (current == tail) {
                    tail = current.prev;
                }
                return;
            }
            current = current.next;
        } while (current != head);
    }

    public void show() {
        if (head == null) {
            return;
        }
        Node current = head;
        do {
            System.out.println("The data address is " + current);
            System.out.println("The data is : " + current.data);
            System.out.println("The next address is : " + current.next);
            System.out.println("The prev address is : " + current.prev);
            System.out.println(" ");
            current = current.next;
        } while (current != head);
    }

    public static void main(String[] args) {
        CircularDoublyLinkedList list = new CircularDoublyLinkedList();
        list.add(10);
        list.add(20);
        list.add(30);
        list.add(40);
        list.delete(10);
        list.show();
    }
}
